File: apps/win-web/src/data/candidate.ts
// CandidateJSON shape — mirrors yunwei_win/services/parse_pipeline/candidate.py.
//
// The frontend ConfirmCard reads this verbatim from the parse pipeline
// output. Adding fields here means adding them on the backend Pydantic
// model first; this file is the consumer.

export type CandidateSourceType = "contract" | "wechat_screenshot" | "excel";

export type CandidateEntityType =
  | "Customer"
  | "Contact"
  | "Contract"
  | "Order"
  | "OrderLine"
  | "OrderItem"
  | "Product"
  | "Invoice"
  | "Payment";

export type CandidateSourceSpan = {
  page?: number | null;
  bbox?: number[] | null;
  text?: string | null;
  cell?: string | null;
  // Adapter-specific extras (sheet name, row index, etc.).
  [extra: string]: unknown;
};

export type CandidateField = {
  name: string;
  value: unknown;
  confidence: number;
  source_span: CandidateSourceSpan;
};

export type CandidateEntity = {
  entity_type: CandidateEntityType;
  temp_id: string;
  fields: CandidateField[];
  missing_required: string[];
};

export type CandidateRelationship = {
  from_temp_id: string;
  to_temp_id: string;
  type: string;
};

export type CandidateSource = {
  type: CandidateSourceType;
  file_ref: string;
  uploaded_by?: string | null;
  uploaded_at?: string | null;
};

export type CandidateJSON = {
  ingestion_id: string;
  source: CandidateSource;
  entities: CandidateEntity[];
  relationships: CandidateRelationship[];
  overall_confidence: number;
  warnings: string[];
};

// ---- Confirm-time edits and submit payload -----------------------------

export type ConfirmFieldDraft = {
  name: string;
  value: unknown;
  confidence: number | null;
  was_edited: boolean;
  source_span: CandidateSourceSpan | null;
};

export type ConfirmEntityDraft = {
  entity_type: CandidateEntityType;
  temp_id: string;
  fields: ConfirmFieldDraft[];
  /** If user picked "associate with existing X" in the duplicate dialog. */
  existing_entity_id?: string | null;
};

export type ConfirmEntitiesRequest = {
  ingestion_id: string;
  source_type: CandidateSourceType | string;
  source_ref: string;
  entities: ConfirmEntityDraft[];
  relationships: CandidateRelationship[];
};

export type WrittenEntity = {
  temp_id: string;
  entity_type: string;
  entity_id: string;
  created: boolean;
  human_verified: boolean;
  verified_by: string;
  field_count: number;
  edited_field_count: number;
};

export type ConfirmEntitiesResponse = {
  written: WrittenEntity[];
  action_log_ids: string[];
};

File: services/platform-api/yunwei_win/services/parse_pipeline/candidate.py
"""Candidate JSON shape.

The strict shape contracted by P0 task ②. Fields can be ADDED for richer
metadata but the documented keys can't be renamed or removed without a
spec bump — task ③ frontend depends on them.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field


EntityType = Literal[
    "Customer",
    "Contact",
    "Contract",
    "Order",
    "OrderLine",  # candidate-JSON / parse-pipeline canonical name
    "OrderItem",  # task ① ontology table name — accepted as alias for forward-compat
    "Product",
    "Invoice",
    "Payment",
]


SourceType = Literal["contract", "wechat_screenshot", "excel"]


class SourceSpan(BaseModel):
    """Pointer back into the input file for one extracted value.

    Visual sources (contract / screenshot) populate ``page`` + ``bbox`` +
    ``text``. Excel populates ``cell`` (e.g. ``"sheet:订单!R7C2"``) and
    sometimes ``text`` (the cell text). ``text`` always carries the
    matched excerpt, regardless of source — task ③ uses it as fallback
    when the visual span can't be rendered.
    """

    model_config = ConfigDict(extra="allow")

    page: int | None = None
    bbox: list[float] | None = None
    text: str | None = None
    cell: str | None = None


class FieldCandidate(BaseModel):
    model_config = ConfigDict(extra="allow")

    name: str
    value: Any | None = None
    confidence: float = 0.0
    source_span: SourceSpan = Field(default_factory=SourceSpan)


class CandidateEntity(BaseModel):
    model_config = ConfigDict(extra="allow")

    entity_type: EntityType
    temp_id: str
    fields: list[FieldCandidate] = Field(default_factory=list)
    missing_required: list[str] = Field(default_factory=list)


class Relationship(BaseModel):
    model_config = ConfigDict(extra="allow")

    from_temp_id: str
    to_temp_id: str
    type: str  # e.g. "Customer-has-Contact", "Order-has-OrderLine"


class SourceInfo(BaseModel):
    model_config = ConfigDict(extra="allow")

    type: SourceType
    file_ref: str
    uploaded_by: str | None = None
    uploaded_at: datetime | None = None


class CandidateJSON(BaseModel):
    model_config = ConfigDict(extra="allow")

    ingestion_id: str = Field(default_factory=lambda: str(uuid4()))
    source: SourceInfo
    entities: list[CandidateEntity] = Field(default_factory=list)
    relationships: list[Relationship] = Field(default_factory=list)
    overall_confidence: float = 0.0
    warnings: list[str] = Field(default_factory=list)
